prymer/variant.py

```diff
@@ -0,0 +1,259 @@
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        """Exit this context manager while closing the underlying VCF handles."""
+        self.close()
+        return None
+
+    def query(self, refname: str, start: int, end: int) -> list[SimpleVariant]:
+        """Fetches variants that overlap a genomic range.
+
+        Args:
+            refname: the reference name
+            start: the 1-based start position
+            end: the 1-based inclusive end position
+        """
+        variants: list[SimpleVariant] = []
+
+        for fh, path in zip(self._readers, self.vcf_paths, strict=True):
+            if not fh.header.contigs.get(refname):
+                _logger.debug(f"Header in VCF file {path} does not contain chromosome {refname}.")
+                continue
+
+            #  pysam.fetch is 0-based, half-open
+            for rec in fh.fetch(contig=refname, start=start - 1, end=end):
+                # Skip over variants with a non-PASS filter
+                filters = list(rec.filter)
+                if len(filters) > 0 and "PASS" not in filters:
+                    continue
+
+                for variant in SimpleVariant.build(rec):
+                    if variant.af is None and self.include_missing_mafs:
+                        variants.append(variant)
+                    elif variant.af is not None and variant.af >= self.min_maf:
+                        variants.append(variant)
+
+        variants.sort(key=lambda v: v.start)
+        return variants
+
+    def mask(self, bases: str, region: Span) -> tuple[str, str]:
+        """
+        Takes a sequence and returns soft and hard masked version of it.  Bases in the sequence
+        are masked if a variant overlaps that base and passes frequency filtering.
+
+        The input sequence is upper-cased, and then masked bases are converted to lower case in the
+        soft-masked version and to Ns in the hard-masked version.
+
+        Args:
+            bases: the bases to be masked
+            region: a span describing the genomic region from which the bases are drawn
+
+        Returns:
+            A tuple of two sequences where the first is soft-masked and the second is hard masked
+        """
+        if len(bases) != region.length:
+            raise ValueError("Region and bases have different lengths: {region} vs. {bases}")
+
+        bases = bases.upper()
+        variants = self.query(refname=region.refname, start=region.start, end=region.end)
+
+        if len(variants) == 0:
+            return bases, bases
+        else:
+            soft = list(bases)
+            hard = list(bases)
+
+            for variant in variants:
+                for pos in range(variant.start, variant.end + 1):
+                    if region.start <= pos <= region.end:
+                        idx = pos - region.start
+                        soft[idx] = soft[idx].lower()
+                        hard[idx] = "N"
```

Comment:
Can we avoid a loop as such?
```suggestion
                start = max(variant.start, region.start)
                end = min(variant.end, region.end) + 1
                soft[start:end] = soft[start:end].lower()
                hard[start:end] = "N" * (end - start + 1)
```

Reply:
Not easily.  `soft[start:end].lower()` doesn't work because `soft[start:end]` is a list slice, not a string.  So that line has to become:

```python
soft[start:end] = [ch.lower() for ch in soft[start:end]]
```

which just moves the loop into a different spot.  I can add in the max/min calls though instead of the `if`.